File: app/src/main/java/com/supercilex/robotscouter/data/util/TeamCache.java
package com.supercilex.robotscouter.data.util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class TeamCache {
    private static final Object TEAM_NAMES_LOCK = new Object();

    private final List<TeamHelper> mTeamHelpers;
    private String mTeamNames;

    public TeamCache(Collection<TeamHelper> teamHelpers) {
        List<TeamHelper> sortedTeamHelpers = new ArrayList<>(teamHelpers);
        Collections.sort(sortedTeamHelpers);
        mTeamHelpers = Collections.unmodifiableList(sortedTeamHelpers);
    }

    public List<TeamHelper> getTeamHelpers() {
        return mTeamHelpers;
    }

    public String getTeamNames() {
        synchronized (TEAM_NAMES_LOCK) {
            if (mTeamNames == null) {
                mTeamNames = TeamHelper.getTeamNames(getTeamHelpers());
            }
            return mTeamNames;
        }
    }
}

File: app/src/main/java/com/supercilex/robotscouter/data/util/TeamCache.kt
package com.supercilex.robotscouter.data.util

import com.supercilex.robotscouter.data.model.Team
import com.supercilex.robotscouter.util.getTeamNames
import java.util.Collections

open class TeamCache(teams: Collection<Team>) {
    val teams: List<Team> by lazy {
        val sortedTeams = ArrayList(teams)
        Collections.sort(sortedTeams)
        Collections.unmodifiableList(sortedTeams)
    }
    val teamNames: String by lazy { getTeamNames(this.teams) }
}
